File: The Last Mile/GAVRP/GAVRP.m
%遗传算法 VRP 问题 Matlab实现 。实现电商订单的分配

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
clear;clc
for A_site=1:124
best_route4 = [];
dis_file=csvread (['A',num2str(A_site),'_dis_AB2.csv']);   %包括原点   site和spot之间的配对
num_file=csvread (['A',num2str(A_site),'_num_AB2.csv'],1,1);   %不包括原点
[m n]=size(dis_file);
num_spot_id=n-1;
Sup_Parent=[1:n-1];
%Sup_Parent=[3 6 7 17 18 5 8 16 19 15 26 25 4 27 29 2 9 14 20 28 30 31 13 22 21 24 11 12 23 10]-1;
G=100;%种群大小
Parent=rand(G,num_spot_id);
%Parent=rand(G,30);%随即父代
for i=1:G
[m n]=sort(Parent(i,:));                     %初始化数据。 对第i行数据排序，返回排序后向量m，和对应的序号n
Parent(i,:)=n;                               %把序号赋给第i行各元素
end                                          %即最后得到100行1-30的随机数组
Pc=0.8;%交叉比率
Pm=0.2;%变异比率
species=[Sup_Parent;Parent];%种群
children=[];%子代
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%g=input('更新时代次数');
h = waitbar(0,'Please wait...');%进度条
for generation=1:floor(num_spot_id*0.5)
%for generation=1:floor(num_spot_id*1.5)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Parent=species;%子代变成父代
children=[];%子代
%选择交叉父代
[n m]=size(Parent);      %返回parent的维数给n和m
% select=rand(1,n)<Pc;
% select=find(select==1);
                                                                                                                %交叉
for i=1:n
    for j=i:n
        if i~=j & rand<Pc             %i不等于j且任一随机数小于PC
            jiaocha
        end
    end
end
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
            




%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
for i=1:n
    if rand<Pm
    parent=Parent(i,:);%变异个体
    X=floor(rand*num_spot_id)+1;
    Y=floor(rand*num_spot_id)+1;
    Z=parent(X);
    parent(X)=parent(Y);
    parent(Y)=Z;                                                                                                 %变异
    children=[children;parent];
    end
end
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%计算子代适应值
[m n]=size(children);
fitness_value_c=zeros(m,1);%子代适应值
for i=1:m
    l1=1;
    for l2=1:n
        if sum(num_file(children(i,l1:l2),1))>135             %135是载重量，保险起见取个比140小的数
            fitness_c
            l1=l2;
        end
        if l2==n
            l2=l2+1;
            fitness_c
        end
                                                                                                                  %计算适应值
        
    
    end
end
%计算父代适应值
[m n]=size(Parent);
fitness_value_P=zeros(m,1);%父代适应值
for i=1:m
    l1=1;
    for l2=1:n
        if sum(num_file(Parent(i,l1:l2),1))>135           % 135是载重量
            fitness_P
            l1=l2;
        end
        if l2==n
            l2=l2+1;
            fitness_P
        end
    end
end
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%淘汰子代
[m n]=sort(fitness_value_c);
children=children(n(1:G),:);   % G种群大小100，保留前100个
fitness_value_c=fitness_value_c(n(1:G));
%淘汰父代
[m n]=sort(fitness_value_P);
Parent=Parent(n(1:G),:);
fitness_value_P=fitness_value_P(n(1:G));
%淘汰种群
species=[children;Parent];
fitness_value=[fitness_value_c;fitness_value_P];
[m n]=sort(fitness_value);
species=species(n(1:G),:);
fitness_value=fitness_value(n(1:G));                                                                             %更新世代
%waitbar(generation/2,h)
waitbar(generation/floor(num_spot_id*0.5),h)

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
end
best_route = species(1,:);%最优线路
best_fit = fitness_value(1);%最优费用
%fitness_value(1:10)
best_route2 = [0];
l1=1;
for l2 = 1:length(best_route)
    if (sum(num_file(best_route(l1:l2)))>135&l2<length(best_route))
        best_route2 = [best_route2,best_route(l1:l2-1),0];
        l1=l2;
    end
    if (sum(num_file(best_route(l1:l2)))<=135&l2==length(best_route))
        best_route2 = [best_route2,best_route(l1:l2),0];
    end
    if (sum(num_file(best_route(l1:l2)))>135&l2==length(best_route))
        best_route2 = [best_route2,best_route(l1:l2-1),0,best_route(l2),0];
        l1=l2;
    end    
end

best_route2;
best_route3 = best_route2+1;
csvwrite(['A',num2str(A_site),'_best_route.csv'],[best_route3';best_fit])
A_site    
end

File: The Last Mile/GAVRP/gavrp.R
# num_all <- list()
# for(i in 1:124){
#   Num <- vector()
#   B <- as.array(AB_group[[i]])
#   for(j in 1:length(B)){
#     if(j==1){
#       Num[j] <- 0
#     }
#     else{
#       Num[j] <- order_1$Num[order_1$Site_id==B[1]&order_1$Spot_id==B[j]]
#     }
#     
#   }
#   num_all_i <- data.frame(B,Num)
#   num_all <- c(num_all,list(num_all_i))
# 
# }
# save(num_all,file = "num_AB.RData")
setwd("E:/workspace/最后一公里")
jiaocha <- function(){
  P1 <- as.numeric(parent[i,]) 
  P2 <- as.numeric(parent[j,]) 
  x <- floor(runif(1,min = 0,max = 1)*(num_spot_id-2))+2
  y <- floor(runif(1,min = 0,max = 1)*(num_spot_id-2))+2
  if(x<y){
    change1 <- P1[x:y]
    change2 <- P2[x:y]
    p1 <- c(P1[(y+1):length(P1)],P1[1:(x-1)],change1)
    p2 <- c(P2[(y+1):length(P2)],P2[1:(x-1)],change2)
    p1 <- p1[!p1%in%change2]
    p2 <- p2[!p2%in%change1]
    P1 <- c(p1[(num_spot_id-y+1):length(p1)],change2,p1[1:(num_spot_id-y)])
    P2 <- c(p2[(num_spot_id-y+1):length(p2)],change1,p2[1:(num_spot_id-y)])
  }
  else{
    change1 <- P1[y:x]
    change2 <- P2[y:x]
    p1 <- c(P1[(x+1):length(P1)],P1[1:(y-1)],change1)
    p2 <- c(P2[(x+1):length(P2)],P2[1:(y-1)],change2)
    p1 <- p1[!p1%in%change2]
    p2 <- p2[!p2%in%change1]
    P1 <- c(p1[(num_spot_id-x+1):length(p1)],change2,p1[1:(num_spot_id-x)])
    P2 <- c(p2[(num_spot_id-x+1):length(p2)],change1,p2[1:(num_spot_id-x)])
  }
  print(i)
  print(j)
  children <- rbind(children,P1,P2)
  return(children)
  
  
}

best_route4 <- list()
best_fit4 <- vector()
load("dis_AB.RData")
load("num_AB.RData")

for(k in 1:1){

  dis_file <- time_all[[k]]
  num_file <- num_all[[k]]
  num_file <- num_file[-1,]
  n <- nrow(dis_file)
  num_spot_id <- n-1
  sup_parent <- c(1:num_spot_id)
  G <- 100
  parent <- data.frame()
  for(i in 1:G){
    parent <- rbind(parent,as.vector(sample(1:num_spot_id,num_spot_id,replace = FALSE)))
  }
  Pc=0.8   #交叉比率
  Pm=0.2   #变异比率
  species <- rbind(sup_parent,parent)
  children <- data.frame()
  print("初始化完成")
  
  # g <- floor(num_spot_id*1.7) #更新迭代次数
  g <- 10
  for(generation in 1:g){
    parent <- species
    children <- data.frame()
    n <- nrow(parent)
    for(i in 1:n){
      for(j in 1:n){
        if(i!=j&runif(1,min = 0,max = 1)<Pc){
          children <- jiaocha()
        }
          
      }
    }
    print("交叉完成")
    for(i in 1:n){
      if(runif(1,min = 0,max = 1)<Pm){
        parent_i <- as.numeric(parent[i,]) 
        x <- floor(runif(1,min = 0,max = 1)*num_spot_id)+1
        y <- floor(runif(1,min = 0,max = 1)*num_spot_id)+1
        z <- parent_i[x]
        parent_i[x] <- parent_i[y]
        parent_i[y] <- z
        children <- rbind(children,parent_i)
      }
    }
    print("变异完成")
    
    m <- nrow(children)
    n <- ncol(children)
    fitness_value_c <- vector()
    for(i in 1:m){
      l1=1
      fitness_value_c_i <- 0
      for(l2 in 1:n){
        if(sum(num_file[as.numeric(children[i,l1:l2]),2])>140){
          path <- as.numeric(children[i,l1:(l2-1)]) 
          l <- length(path)
          for(k in 1:l){
            if(k==1){
              fitness_value_c_i <- fitness_value_c_i+3*sqrt(num_file[path[1],2])+5+dis_file[1,path[1]+1]
            }
            else{
              fitness_value_c_i <- fitness_value_c_i+3*sqrt(num_file[path[k],2])+5+dis_file[path[k-1]+1,path[k]+1]
            }
          }
          fitness_value_c_i <- fitness_value_c_i+dis_file[1,path[l]+1]
          l1 <- l2
        }
        if(l2==n){
          l2 <- l2+1
          path <- as.numeric(children[i,l1:(l2-1)])
          l <- length(path)
          for(k in 1:l){
            if(k==1){
              fitness_value_c_i <- fitness_value_c_i+3*sqrt(num_file[path[1],2])+5+dis_file[1,path[1]+1]
            }
            else{
              fitness_value_c_i <- fitness_value_c_i+3*sqrt(num_file[path[k],2])+5+dis_file[path[k-1]+1,path[k]+1]
            }
          }
          fitness_value_c_i <- fitness_value_c_i+dis_file[1,path[l]+1]
          
        }
      }
      fitness_value_c <- c(fitness_value_c,fitness_value_c_i)
    }
    print("子代适应值")
    
    m <- nrow(parent)
    n <- ncol(parent)
    fitness_value_p <- vector()
    for(i in 1:m){
      l1=1
      fitness_value_p_i <- 0
      for(l2 in 1:n){
        if(sum(num_file[as.numeric(parent[i,l1:l2]),2] )>140){
          path <- as.numeric(parent[i,l1:(l2-1)]) 
          l <- length(path)
          for(k in 1:l){
            if(k==1){
              fitness_value_p_i <- fitness_value_p_i+3*sqrt(num_file[path[1],2])+5+dis_file[1,path[1]+1]
            }
            else{
              fitness_value_p_i <- fitness_value_p_i+3*sqrt(num_file[path[k],2])+5+dis_file[path[k-1]+1,path[k]+1]
            }
          }
          fitness_value_p_i <- fitness_value_p_i+dis_file[1,path[l]+1]
          l1 <- l2
        }
        if(l2==n){
          l2 <- l2+1
          path <- as.numeric(parent[i,l1:(l2-1)])
          l <- length(path)
          for(k in 1:l){
            if(k==1){
              fitness_value_p_i <- fitness_value_p_i+3*sqrt(num_file[path[1],2])+5+dis_file[1,path[1]+1]
            }
            else{
              fitness_value_p_i <- fitness_value_p_i+3*sqrt(num_file[path[k],2])+5+dis_file[path[k-1]+1,path[k]+1]
            }
          }
          fitness_value_p_i <- fitness_value_p_i+dis_file[1,path[l]+1]
          
        }
      }
      fitness_value_p <- c(fitness_value_p,fitness_value_p_i)
    }
    print("父代适应值")
    
    n <- sort(fitness_value_c,index.return=TRUE)$ix
    children <- children[n[1:G],]
    fitness_value_c <- fitness_value_c[n[1:G]]
    n <- sort(fitness_value_p,index.return=TRUE)$ix
    parent <- parent[n[1:G],]
    fitness_value_p <- fitness_value_p[n[1:G]]
    colnames(children) <- colnames(parent)
    species <- rbind(children,parent)
    fitness_value <- c(fitness_value_c,fitness_value_p)
    n <- sort(fitness_value,index.return=TRUE)$ix
    species <- species[n[1:G],]
    fitness_value <- fitness_value[n[1:G]]
  }
  print("迭代完成")
  best_route <- as.numeric(species[1,]) 
  best_fit <- fitness_value[1]
  best_route2 <- 0
  l1 <- 1
  for(l2 in 1:length(best_route)){
    if(sum(num_file[best_route[l1:l2],2])>140&l2<length(best_route)){
      best_route2 <- c(best_route2,best_route[l1:(l2-1)],0)
      l1 = l2
    }
    if(sum(num_file[best_route[l1:l2],2])<=140&l2==length(best_route)){
      best_route2 <- c(best_route2,best_route[l1:l2],0)
    }
    if(sum(num_file[best_route[l1:l2],2])>140&l2==length(best_route)){
      best_route2 <- c(best_route2,best_route[l1:(l2-1)],0,best_route[l2],0)
      l1 = l2
    }
  }
  best_route3 <- best_route2 +1
  best_fit2 <- 0
  for(l in 1:(length(best_route2)-1)){
    best_fit2 <- best_fit2+dis_file[best_route3[l],best_route3[l+1]]
  }
  # print(best_fit2)
#   print(best_route3)
#   print(best_fit)
#   print(k)
#   best_route4 <- c(best_route4,list(best_route3))
#   best_fit4 <- c(best_fit4,best_fit)
  

}







  